sway-lsp/src/capabilities/on_enter.rs

```diff
@@ -0,0 +1,203 @@
+use crate::{
+    config::OnEnterConfig,
+    core::{document::TextDocument, session::Session},
+};
+use std::sync::Arc;
+use tower_lsp::{
+    lsp_types::{
+        DidChangeTextDocumentParams, DocumentChanges, OneOf,
+        OptionalVersionedTextDocumentIdentifier, Position, Range, TextDocumentEdit, TextEdit, Url,
+        WorkspaceEdit,
+    },
+    Client,
+};
+
+const NEWLINE: &str = "\n";
+const COMMENT_START: &str = "//";
+const DOC_COMMENT_START: &str = "///";
+
+/// If the change was an enter keypress or pasting multiple lines in a comment, it prefixes the line(s)
+/// with the appropriate comment start pattern (// or ///).
+pub(crate) async fn on_enter(
+    config: &OnEnterConfig,
+    client: &Client,
+    session: &Arc<Session>,
+    temp_uri: &Url,
+    params: &DidChangeTextDocumentParams,
+) {
+    if !(params.content_changes[0].text.contains(NEWLINE)) {
+        return;
+    }
+
+    let mut workspace_edit = None;
+    let text_document = session
+        .get_text_document(temp_uri)
+        .expect("could not get text document");
+
+    if config.continue_doc_comments.unwrap_or(false) {
+        workspace_edit = get_comment_workspace_edit(DOC_COMMENT_START, params, &text_document);
+    }
+
+    if config.continue_comments.unwrap_or(false) && workspace_edit.is_none() {
+        workspace_edit = get_comment_workspace_edit(COMMENT_START, params, &text_document);
+    }
+
+    // Apply any edits.
+    if let Some(edit) = workspace_edit {
+        if let Err(err) = client.apply_edit(edit).await {
+            tracing::error!("on_enter failed to apply edit: {}", err);
+        }
+    }
+}
+
+fn get_comment_workspace_edit(
+    start_pattern: &str,
+    change_params: &DidChangeTextDocumentParams,
+    text_document: &TextDocument,
+) -> Option<WorkspaceEdit> {
+    let range = change_params.content_changes[0]
+        .range
+        .expect("change is missing range");
+    let line = text_document.get_line(range.start.line.try_into().expect("failed to get line"));
```

Comment:
Is it possible to return an error instead of panicking? We don't want to crash the server in the situation where the line couldn't be retrieved.

Reply:
changed to casting